File: src/ai_multi_reference_timekeeping/observability.py
"""Metrics export and health checks."""

from __future__ import annotations

from dataclasses import dataclass
import json
import threading
import time
from http.server import BaseHTTPRequestHandler, HTTPServer
from typing import Iterable

from .metrics import mtie, tdev


@dataclass
class HealthStatus:
    """Health status snapshot."""

    last_update: float
    shm_last_write: float | None
    ok: bool


class HealthMonitor:
    """Track freshness of TimeServer updates and SHM writes."""

    def __init__(self, freshness_window: float = 10.0) -> None:
        self._freshness_window = freshness_window
        self._last_update: float | None = None
        self._shm_last_write: float | None = None

    def mark_update(self) -> None:
        self._last_update = time.time()

    def mark_shm_write(self) -> None:
        self._shm_last_write = time.time()

    def status(self) -> HealthStatus:
        now = time.time()
        last_update = self._last_update or 0.0
        ok = now - last_update <= self._freshness_window if last_update else False
        return HealthStatus(last_update=last_update, shm_last_write=self._shm_last_write, ok=ok)


class MetricsTracker:
    """Maintain a rolling buffer of offsets for metrics export."""

    def __init__(self, window_size: int = 60) -> None:
        self._window_size = window_size
        self._offsets: list[float] = []

    def update(self, offset: float) -> None:
        self._offsets.append(offset)
        if len(self._offsets) > self._window_size:
            self._offsets.pop(0)

    def metrics(self) -> dict[str, float]:
        if len(self._offsets) < 3:
            return {}
        return {
            "tdev_tau1": tdev(self._offsets, tau=1),
            "mtie_window2": mtie(self._offsets, window=2),
        }


class MetricsExporter:
    """HTTP server exposing metrics and health endpoints."""

    def __init__(self, tracker: MetricsTracker, health: HealthMonitor) -> None:
        self._tracker = tracker
        self._health = health
        self._server: HTTPServer | None = None
        self._thread: threading.Thread | None = None

    def start(self, host: str, port: int) -> None:
        tracker = self._tracker
        health = self._health

        class Handler(BaseHTTPRequestHandler):
            def do_GET(self) -> None:  # noqa: N802
                if self.path == "/metrics":
                    body = []
                    for key, value in tracker.metrics().items():
                        body.append(f"{key} {value}")
                    response = "\n".join(body).encode()
                    self.send_response(200)
                    self.send_header("Content-Type", "text/plain")
                    self.end_headers()
                    self.wfile.write(response)
                elif self.path == "/health":
                    status = health.status()
                    response = json.dumps(status.__dict__).encode()
                    self.send_response(200 if status.ok else 503)
                    self.send_header("Content-Type", "application/json")
                    self.end_headers()
                    self.wfile.write(response)
                else:
                    self.send_response(404)
                    self.end_headers()

            def log_message(self, format: str, *args: Iterable[object]) -> None:  # noqa: A003
                return

        self._server = HTTPServer((host, port), Handler)
        self._thread = threading.Thread(target=self._server.serve_forever, daemon=True)
        self._thread.start()

    def stop(self) -> None:
        if self._server:
            self._server.shutdown()
            self._server.server_close()
        if self._thread:
            self._thread.join(timeout=1.0)

File: cpp/src/observability.cpp
#include "aimrt/observability.hpp"

#include <arpa/inet.h>
#include <netinet/in.h>
#include <sys/socket.h>
#include <unistd.h>

#include <cstring>
#include <sstream>

#include "aimrt/common.hpp"
#include "aimrt/metrics.hpp"

namespace aimrt {

HealthMonitor::HealthMonitor(double freshness_window) : freshness_window_(freshness_window) {}

void HealthMonitor::mark_update() {
    std::lock_guard<std::mutex> guard(mutex_);
    last_update_ = seconds_since_epoch();
}

void HealthMonitor::mark_shm_write() {
    std::lock_guard<std::mutex> guard(mutex_);
    shm_last_write_ = seconds_since_epoch();
}

HealthStatus HealthMonitor::status() const {
    std::lock_guard<std::mutex> guard(mutex_);
    const double now = seconds_since_epoch();
    const double last_update = last_update_.value_or(0.0);
    const bool ok = last_update_.has_value() && (now - last_update <= freshness_window_);
    return HealthStatus{last_update, shm_last_write_, ok};
}

MetricsTracker::MetricsTracker(int window_size) : window_size_(window_size) {}

void MetricsTracker::update(double offset) {
    offsets_.push_back(offset);
    if (static_cast<int>(offsets_.size()) > window_size_) {
        offsets_.erase(offsets_.begin());
    }
}

std::vector<std::pair<std::string, double>> MetricsTracker::metrics() const {
    if (offsets_.size() < 3) {
        return {};
    }
    std::vector<std::pair<std::string, double>> output;
    output.emplace_back("tdev_tau1", tdev(offsets_, 1));
    output.emplace_back("mtie_window2", mtie(offsets_, 2));
    return output;
}

MetricsExporter::MetricsExporter(MetricsTracker& tracker, HealthMonitor& health)
    : tracker_(tracker), health_(health) {}

MetricsExporter::~MetricsExporter() {
    stop();
}

void MetricsExporter::start(const std::string& host, int port) {
    if (running_.exchange(true)) {
        return;
    }
    thread_ = std::thread(&MetricsExporter::serve, this, host, port);
}

void MetricsExporter::stop() {
    if (!running_.exchange(false)) {
        return;
    }
    if (server_fd_ >= 0) {
        ::shutdown(server_fd_, SHUT_RDWR);
    }
    if (thread_.joinable()) {
        thread_.join();
    }
}

void MetricsExporter::serve(const std::string& host, int port) {
    server_fd_ = ::socket(AF_INET, SOCK_STREAM, 0);
    if (server_fd_ < 0) {
        return;
    }

    int opt = 1;
    ::setsockopt(server_fd_, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt));

    sockaddr_in addr{};
    addr.sin_family = AF_INET;
    addr.sin_port = htons(static_cast<uint16_t>(port));
    ::inet_pton(AF_INET, host.c_str(), &addr.sin_addr);

    if (::bind(server_fd_, reinterpret_cast<sockaddr*>(&addr), sizeof(addr)) < 0) {
        ::close(server_fd_);
        server_fd_ = -1;
        return;
    }
    if (::listen(server_fd_, 4) < 0) {
        ::close(server_fd_);
        server_fd_ = -1;
        return;
    }

    while (running_) {
        sockaddr_in client{};
        socklen_t len = sizeof(client);
        int client_fd = ::accept(server_fd_, reinterpret_cast<sockaddr*>(&client), &len);
        if (client_fd < 0) {
            continue;
        }

        char buffer[1024] = {0};
        const ssize_t read_bytes = ::read(client_fd, buffer, sizeof(buffer) - 1);
        if (read_bytes <= 0) {
            ::close(client_fd);
            continue;
        }

        std::string request(buffer, static_cast<size_t>(read_bytes));
        auto first_line_end = request.find("\r\n");
        std::string first_line = first_line_end == std::string::npos ? request : request.substr(0, first_line_end);
        auto parts = split(first_line, ' ');
        std::string path = parts.size() >= 2 ? parts[1] : "/";

        std::string body;
        std::string status = "200 OK";
        std::string content_type = "text/plain";

        if (path == "/metrics") {
            std::ostringstream response;
            for (const auto& [key, value] : tracker_.metrics()) {
                response << key << " " << value << "\n";
            }
            body = response.str();
        } else if (path == "/health") {
            auto status_info = health_.status();
            std::ostringstream response;
            response << "{\"last_update\":" << status_info.last_update
                     << ",\"ok\":" << (status_info.ok ? "true" : "false");
            if (status_info.shm_last_write.has_value()) {
                response << ",\"shm_last_write\":" << status_info.shm_last_write.value();
            }
            response << "}";
            body = response.str();
            content_type = "application/json";
            if (!status_info.ok) {
                status = "503 Service Unavailable";
            }
        } else {
            status = "404 Not Found";
            body = "";
        }

        std::ostringstream header;
        header << "HTTP/1.1 " << status << "\r\n"
               << "Content-Type: " << content_type << "\r\n"
               << "Content-Length: " << body.size() << "\r\n\r\n";

        const std::string response = header.str() + body;
        ::send(client_fd, response.c_str(), response.size(), 0);
        ::close(client_fd);
    }

    if (server_fd_ >= 0) {
        ::close(server_fd_);
        server_fd_ = -1;
    }
}

}  // namespace aimrt
